Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.h

```diff
@@ -0,0 +1,98 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibJS/Runtime/Object.h>
+#include <LibWasm/AbstractMachine/AbstractMachine.h>
+#include <LibWeb/Forward.h>
+
+namespace Web::Bindings {
+
+class WebAssemblyObject final : public JS::Object {
+    JS_OBJECT(WebAssemblyObject, JS::Object);
+
+public:
+    explicit WebAssemblyObject(JS::GlobalObject&);
+    virtual void initialize(JS::GlobalObject&) override;
+    virtual ~WebAssemblyObject() override = default;
+
+    struct CompiledWebAssemblyModule {
+        explicit CompiledWebAssemblyModule(Wasm::Module&& module)
+            : module(move(module))
+        {
+        }
+
+        Wasm::Module module;
+    };
+
+    // FIXME: These should just be members of the module (instance) object,
+    //        but the module needs to stick around while its instance is alive
+    //        so ideally this would be a refcounted object, shared between
+    //        WebAssemblyModuleObject's and WebAssemblyInstantiatedModuleObject's.
+    static NonnullOwnPtrVector<CompiledWebAssemblyModule> s_compiled_modules;
+    static NonnullOwnPtrVector<Wasm::ModuleInstance> s_instantiated_modules;
+
+    static Wasm::AbstractMachine s_abstract_machine;
+
+private:
+    JS_DECLARE_NATIVE_FUNCTION(validate);
+    JS_DECLARE_NATIVE_FUNCTION(compile);
+    JS_DECLARE_NATIVE_FUNCTION(instantiate);
+};
+
+class WebAssemblyModuleObject final : public JS::Object {
+    JS_OBJECT(WebAssemblyModuleObject, JS::Object);
+
+public:
+    explicit WebAssemblyModuleObject(JS::GlobalObject&, size_t index);
+    virtual ~WebAssemblyModuleObject() override = default;
+
+    size_t index() const { return m_index; }
+    const Wasm::Module& module() const { return WebAssemblyObject::s_compiled_modules.at(m_index).module; }
+
+private:
+    size_t m_index { 0 };
+};
+
+class WebAssemblyModuleInstanceObject final : public JS::Object {
+    JS_OBJECT(WebAssemblyModuleInstanceObject, JS::Object);
+
+public:
+    explicit WebAssemblyModuleInstanceObject(JS::GlobalObject&, size_t index);
+    virtual void initialize(JS::GlobalObject&) override;
+    virtual ~WebAssemblyModuleInstanceObject() override = default;
+
+    size_t index() const { return m_index; }
+    Wasm::ModuleInstance& instance() const { return WebAssemblyObject::s_instantiated_modules.at(m_index); }
+
+    void visit_edges(Cell::Visitor&) override;
+
+private:
+    JS_DECLARE_NATIVE_GETTER(exports_getter);
+
+    size_t m_index { 0 };
+    JS::Object* m_exports_object { nullptr };
+};
+
+class WebAssemblyMemoryObject final : public JS::Object {
+    JS_OBJECT(WebAssemblyModuleObject, JS::Object);
+
+public:
+    explicit WebAssemblyMemoryObject(JS::GlobalObject&, Wasm::MemoryAddress);
+    virtual void initialize(JS::GlobalObject&) override;
+    virtual ~WebAssemblyMemoryObject() override = default;
+
+    auto address() const { return m_address; }
+
+private:
+    // JS_DECLARE_NATIVE_FUNCTION(grow);
+    JS_DECLARE_NATIVE_GETTER(buffer);
```

Comment:
Totally forgot this previously, this needs to have a distinct prototype (not `global_object.object_prototype()`) and not be on the instance directly (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Instance/exports) - sorry!

This should be relatively easy, but if you need any help just let me know :)

Reply:
Oh I see (I assume you meant to comment on `JS_DECLARE_NATIVE_GETTER(exports_getter)`)
so I should make a `WebAssemblyInstancePrototype` and put these on that instead I guess?

Also this comment is annoying, I'll get rid of it or implement it

Reply:
> I assume you meant to comment on JS_DECLARE_NATIVE_GETTER(exports_getter)

oops, yes!

> so I should make a WebAssemblyInstancePrototype and put these on that instead I guess?

Precisely 👍

Reply:
Hmmm
How do I make all of them share one prototype object?

Reply:
Basically like this:

https://github.com/SerenityOS/serenity/blob/bdd7741ae12955bd31225870fa85e96a6e79e9dd/Userland/Libraries/LibWeb/Bindings/WindowObjectHelper.h#L227-L233

https://github.com/SerenityOS/serenity/blob/1682f0b760c736e7bbb62bea86248bd83579bec7/Userland/Libraries/LibWeb/Bindings/WindowObject.h#L35-L56

Which allows you to get the unique prototype like this:

https://github.com/SerenityOS/serenity/blob/bdd7741ae12955bd31225870fa85e96a6e79e9dd/Userland/Libraries/LibWeb/CodeGenerators/WrapperGenerator.cpp#L904

TL;DR: Create a prototype once, add it to the global object's `m_prototypes` to save it from GC, set it on `WebAssemblyInstanceObject`. You don't have to use these specific functions, they're mostly for the generated wrapper code.

Reply:
I can't just insert random stuff into the global object though, this is inserted from LibWeb, and I don't have access to the WindowObject directly (I can probably cast the global object to that), but that's just ugly...

Reply:
`static_cast<WindowObject&>(global_object)` is exactly how we do this :)